File: base/science-tech-maths/programming/algorithms/job-interviews/snippets/factorial.cpp
#include <iostream>

unsigned int factorial_rec(unsigned int n) {
    if (n < 0) {
        throw;
    }
    return n > 1 ? n * factorial_rec(n - 1) : 1;
}

unsigned int factorial_iter(unsigned int n) {
    if (n < 0) {
        throw;
    }

    unsigned int fact(1);
    for (unsigned int i = 1; i < n + 1; ++i) {
        fact *= i;
    }
    return fact;
}

int main(int argc, const char** argv) {
    std::cout << factorial_rec(0) << std::endl;
    std::cout << factorial_rec(1) << std::endl;
    std::cout << factorial_rec(4) << std::endl;   // 2 3 4
    std::cout << factorial_rec(10) << std::endl;  // 3628800
    std::cout << factorial_iter(0) << std::endl;
    std::cout << factorial_iter(1) << std::endl;
    std::cout << factorial_iter(4) << std::endl;   // 2 3 4
    std::cout << factorial_iter(10) << std::endl;  // 3628800
    return 0;
}

File: base/science-tech-maths/programming/algorithms/job-interviews/snippets/factorial.py
# O(N)
def factorial_iter(n):
    fact = 1
    for i in range(1, n + 1):
        fact *= i
    return fact


# even if it's recursive, the complexity is still O(N)
def factorial_rec(n):
    if n < 0:
        return -1
    elif n == 0:
        return 1
    else:
        return n * factorial_rec(n - 1)


if __name__ == "__main__":
    print(factorial_iter(4) == 24)
    print(factorial_rec(4) == 24)
    print(factorial_iter(6) == 720)
    print(factorial_rec(6) == 720)
